src/Altinn.AccessManagement.Core/Services/DelegationsService.cs

```diff
@@ -20,80 +23,120 @@ public class DelegationsService : IDelegationsService
-            if (keyrolePartyIds.Any())
+            Party toParty = await _contextRetrievalService.GetPartyAsync(partyId);
+            if (toParty == null || toParty.PartyTypeName != PartyType.Organisation)
+            {
+                throw new ValidationException($"Maskinporten schema delegation can only be delegated To a valid PartyId belonging to an organization. Invalid value: {partyId}");
+            }
+
+            // Verify authenticated users delegable rights
+            RightsQuery rightsQuery = new RightsQuery
+            {
+                To = new List<AttributeMatch> { new AttributeMatch { Id = AltinnXacmlConstants.MatchAttributeIdentifiers.UserAttribute, Value = delegatingUserId.ToString() } },
+                From = new List<AttributeMatch> { new AttributeMatch { Id = AltinnXacmlConstants.MatchAttributeIdentifiers.PartyAttribute, Value = from } },
+                Resource = right.Resource
+            };
+            List<Right> usersDelegableRights = await _pip.GetRights(rightsQuery, getDelegableRights: true);
+            if (usersDelegableRights.IsNullOrEmpty())
             {
-                List<DelegationChange> keyRoleDelegations = resourceMatchType == ResourceAttributeMatchType.AltinnAppId ?
-                    await _delegationRepository.GetAllCurrentAppDelegationChanges(offeredByPartyIds, resourceIds, coveredByPartyIds: keyrolePartyIds) :
-                    await _delegationRepository.GetAllCurrentResourceRegistryDelegationChanges(offeredByPartyIds, resourceIds, coveredByPartyIds: keyrolePartyIds);
+                throw new ValidationException($"Authenticated user does not have any delegable rights for the resource: {resourceRegistryId}");
+            }
+
+            if (usersDelegableRights.Any(r => !r.CanDelegate.Value)) 
+            {
+                // ToDo: include SecurityLevelInfo on delegable rights
+                throw new ValidationException($"Authenticated user does not meet the required security level requirement for resource: {resourceRegistryId}. Current: N/A Required: N/A");
+            }
 
-                if (keyRoleDelegations.Any())
+            // Perform delegation
+            List<Rule> rulesToDelegate = new List<Rule>();
+            foreach (Right rightToDelegate in usersDelegableRights)
+            {
+                rulesToDelegate.Add(new Rule
                 {
-                    delegations.AddRange(keyRoleDelegations);
-                }
+                    DelegatedByUserId = delegatingUserId,
+                    OfferedByPartyId = int.Parse(from),
+                    CoveredBy = delegation.To,
+                    Resource = rightToDelegate.Resource,
+                    Action = rightToDelegate.Action
+                });
             }
 
-            return delegations;
+            List<Rule> result = await _pap.TryWriteDelegationPolicyRules(rulesToDelegate);
+            if (result.All(r => r.CreatedSuccessfully))
```

Comment:
Must either handle partial success of delegation or be able to tell _pap.TryWriteDelegationPolicyRules to not allow partial success.

Reply:
Since all rules is for the same resource, this should theoretically "never" happen, unless the policy is somehow different/updated between when the delegable rights of the user has been retrieved and the execution in _pap.TryWriteDelegationPolicyRules.